### PR TITLE
✨ Initial experimental Azure DevOps client

### PR DESCRIPTION
**Pull request by JamieMagee**
#### What kind of change does this PR introduce?

This PR introduces early experimental support for Azure DevOps. It's the MVP for a scorecard scan to complete without panicing.

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

No Support for Azure DevOps repository scanning

#### What is the new behavior (if this is a feature change)?**

This is a followup from #4178, and the next step for #4177. It is the minimal client implementation for Azure DevOps repository support.

I'd like to get feedback at this stage, before I implement the remainder of the client.

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Continued work on #4177 

#### Special notes for your reviewer

This change is currenltly gated behind the `SCORECARD_EXPERIMENTAL` environment variable, same as the initial GitLab client was.

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
Initial experimental Azure DevOps client
```

### CODE REVIEW

File: clients/azuredevopsrepo/client.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/client.go:194]: So the GitHub client doesn't need a repo because it's always `github.com` (unless you set a specific env var), the GitLab client takes a host.

It seems Azure needs an `organizationUrl`. Do you have thoughts on accepting that as a parameter instead of a repo? Or does this give you more flexibility in the future?

File: clients/azuredevopsrepo/client.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/client.go:215]: This isn't part of the `RepoClient` interface, let's just delete this. 

Plus oss fuzz supports arbitrary git hosts

File: clients/azuredevopsrepo/commits.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/commits.go:56]: this is the <commitDepth> most recent?

File: clients/azuredevopsrepo/commits.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/commits.go:60]: note: Scorecard also has a `--commit` argument, which you have stored in the `handler.repourl.commitSHA`. So we should respect that

File: clients/azuredevopsrepo/commits.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/commits.go:65]: same comment about context reuse, as there is a top level ctx somewhere we can use

File: cmd/root.go
Comment (spencerschrock) [inline cmd/root.go:227]: experimental gate here?

Your other gate was in `checker.GetClients`, which isn't used by most of the code

File: cmd/root.go
Comment (spencerschrock) [inline cmd/root.go:229]:
```go
errors.Join(errGitHub, errGitLab, errAzureDevOps)
```

**Review by spencerschrock — COMMENTED** [review]
> It's the MVP for a scorecard scan to complete without panicing.

panicing, or completing without runtime errors?  What's the manual testing of this look like for the maintainers? Are there any world-public azuredevops repos?

> I'd like to get feedback at this stage, before I implement the remainder of the client.

In general, the pattern is good. You copied the existing structure, and you're working through it piece by piece. 

Did you plan on finishing the client in this PR or followups? Breaking it up in separate PRs may be easier. Your first approach was "run scorecard without panicing", but I think a good next step is "run a specific check"

So start with something like `go run main.go --repo <blah> --checks Maintained`, which just needs some info about the repo, commits, and issues. Code-Review is probably in this group too.

Then you can move on to checks which need file content (Binary-Artifacts may be a good one?), and then some of the more involved ones that rely on CI/CD data etc.

**Comment by spencerschrock** [comment]
There's a lot of linter errors that I'm sure you're waiting to resolve until after your approach got reviewed. So I didn't review any of that sort of thing, happy to give tips if desired.

File: clients/azuredevopsrepo/client.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/client.go:46]: I'll follow your lead and change this to match Go conventions. I don't usually write in Go so I appreciate the help here.

File: clients/azuredevopsrepo/repo.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/repo.go:15]: Does it matter that this was dead code before?

File: clients/azuredevopsrepo/client.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/client.go:215]: Removed.

File: clients/azuredevopsrepo/client.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/client.go:194]: I do need an `organizationUrl`, but I use `repo` here to add support for on-premises Azure DevOps instances. The host is always `dev.azure.com` for hosted Azure DevOps, but it can be anything for on-premises.

File: clients/azuredevopsrepo/commits.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/commits.go:56]: Yes, commits are returned in reverse chronological order: https://learn.microsoft.com/en-us/rest/api/azure/devops/git/commits/get-commits?view=azure-devops-rest-7.1&tabs=HTTP

File: clients/azuredevopsrepo/client.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/client.go:94]: Rate limits are based around something called Azure DevOps Throughput Units AKA TSTUs (The TS used to stand for Team Services). TSTUs encapsulate computational, database, and bandwith cost for a request. You get 200 TSTUs in a sliding 5 minute window. The documentation is here: [Rate and usage limits][1]

This call to `GetRepository` costs ~0.1 TSTUs, and in general I'd be very surprised if a single run of scorecard would hit a rate limit. But Azure Devops returns `Retry-After`, `X-RateLimit-Remaining`, and a few other headers to help clients.

Public repositories do exist, but it's not common for Azure DevOps. I've been testing using my own organization: https://dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee

[1]: https://learn.microsoft.com/en-us/azure/devops/integrate/concepts/rate-limits?view=azure-devops

**Comment by JamieMagee** [comment]
> panicing, or completing without runtime errors? What's the manual testing of this look like for the maintainers? Are there any world-public azuredevops repos?

Completing without runtime errors. Here's the output, running against https://dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee, after my most recent changes:

<details>
<summary>Details</summary>

```
Aggregate score: 0.0 / 10

Check scores:
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| SCORE  |          NAME          |                REASON                 |                             DOCUMENTATION/REMEDIATION                             |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Binary-Artifacts       | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#binary-artifacts       |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Branch-Protection      | internal error: unsupported           | https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection      |
|        |                        | feature                               |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | CI-Tests               | no pull request found                 | https://github.com/ossf/scorecard/blob/main/docs/checks.md#ci-tests               |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10 | CII-Best-Practices     | no effort to earn an OpenSSF          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#cii-best-practices     |
|        |                        | best practices badge detected         |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10 | Code-Review            | Found 0/1 approved changesets         | https://github.com/ossf/scorecard/blob/main/docs/checks.md#code-review            |
|        |                        | -- score normalized to 0              |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Contributors           | internal error:                       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#contributors           |
|        |                        | Client.Repositories.ListContributors: |                                                                                   |
|        |                        | unsupported feature                   |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Dangerous-Workflow     | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow     |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Dependency-Update-Tool | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dependency-update-tool |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Fuzzing                | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#fuzzing                |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | License                | internal error:                       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#license                |
|        |                        | fileparser.OnAllFilesDo: error        |                                                                                   |
|        |                        | during ListFiles: unsupported         |                                                                                   |
|        |                        | feature                               |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Maintained             | internal error: unsupported           | https://github.com/ossf/scorecard/blob/main/docs/checks.md#maintained             |
|        |                        | feature                               |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Packaging              | packaging workflow not                | https://github.com/ossf/scorecard/blob/main/docs/checks.md#packaging              |
|        |                        | detected                              |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Pinned-Dependencies    | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies    |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | SAST                   | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#sast                   |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | SBOM                   | internal error:                       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#sbom                   |
|        |                        | RepoClient.ListReleases:              |                                                                                   |
|        |                        | unsupported feature                   |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Security-Policy        | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#security-policy        |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Signed-Releases        | internal error: unsupported           | https://github.com/ossf/scorecard/blob/main/docs/checks.md#signed-releases        |
|        |                        | feature                               |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Token-Permissions      | internal error: error during          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions      |
|        |                        | ListFiles: unsupported feature        |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Vulnerabilities        | internal error:                       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#vulnerabilities        |
|        |                        | RepoClient.LocalPath:                 |                                                                                   |
|        |                        | unsupported feature                   |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
| ?      | Webhooks               | internal error: internal error:       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#webhooks               |
|        |                        | Client.Repositories.ListWebhooks      |                                                                                   |
|--------|------------------------|---------------------------------------|-----------------------------------------------------------------------------------|
```

</details>

> Did you plan on finishing the client in this PR or followups? Breaking it up in separate PRs may be easier. Your first approach was "run scorecard without panicing", but I think a good next step is "run a specific check"
> 
> So start with something like `go run main.go --repo <blah> --checks Maintained`, which just needs some info about the repo, commits, and issues. Code-Review is probably in this group too.
> 
> Then you can move on to checks which need file content (Binary-Artifacts may be a good one?), and then some of the more involved ones that rely on CI/CD data etc.

I think I'd like to finish the client in a followup PR if that's possible? I think it'd be easier for you to review in smaller chunks instead of one, big bang, pull request.

I plan to attend the scorecard APAC meeting on Thursday to gauge interest as well.

> There's a lot of linter errors that I'm sure you're waiting to resolve until after your approach got reviewed. So I didn't review any of that sort of thing, happy to give tips if desired.

I'll tackle those in the next day or two.

Thanks so much for the review!

File: clients/azuredevopsrepo/repo.go
Comment (spencerschrock) [reply clients/azuredevopsrepo/repo.go:15]: In a technical sense, no, as parts of the codebase are as a library. So a dev could have written code which depends on "github.com/ossf/scorecard/v5/clients/azuredevops", and renaming this would be a breaking change for them.

But the `azuredevops` package wasn't merged before the v5.0.0 release, so the only way someone could have depended on this was with a [pseudo-version](https://go.dev/ref/mod#pseudo-versions). I'm not sure what the official stance is on pseudo-versions and semver, but I don't consider this breaking anymore.

File: clients/azuredevopsrepo/commits.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/commits.go:144]: you should just be able to do:
```go
commits[i].AssociatedMergeRequest = associatedPullRequest
```

File: clients/azuredevopsrepo/repo.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/repo.go:32]: `ID` should be `id` to hide the type.

**Review by spencerschrock — APPROVED** [review]
> I think I'd like to finish the client in a followup PR if that's possible? I think it'd be easier for you to review in smaller chunks

Agreed. I went ahead and approved the boilerplate, pending linter fixes.

>> There's a lot of linter errors that I'm sure you're waiting to resolve until after your approach got reviewed. So I didn't review any of that sort of thing, happy to give tips if desired.
>
> I'll tackle those in the next day or two.

`make fix-linter` may help with some auto-generated fixes.
Anything that complains about struct alignment you can usually fix with 
```
go install golang.org/x/tools/go/analysis/passes/fieldalignment/cmd/fieldalignment
fieldalignment -fix ./clients/azuredevopsrepo
```

Happy to help with any others as well.

**Comment by spencerschrock** [comment]
also your latest commit wasn't DCO'd. that will also be a blocker.

File: clients/azuredevopsrepo/client.go
Comment (spencerschrock) [reply clients/azuredevopsrepo/client.go]: Discussed in the biweekly community meeting: The Scorecard maintainers should take a look into a test organization so we can make some test repos and get a PAT to do e2e tests later

File: clients/azuredevopsrepo/client.go
Comment (JamieMagee) [reply clients/azuredevopsrepo/client.go:57]: I think I've got the syntax for this right, but let me know if not.

**Comment by JamieMagee** [comment]
@spencerschrock I think I've addressed all your comments. If so, this should be ready to merge.

**Comment by spencerschrock** [comment]
> @spencerschrock I think I've addressed all your comments. If so, this should be ready to merge.

I've been out on the office on personal time, will take a look in the next day or two.

File: clients/azuredevopsrepo/client.go
Comment (spencerschrock) [reply clients/azuredevopsrepo/client.go:61]: Both `InitRepo` and `IsArchived` make a call to `c.azdoClient.GetRepository`, so I was imagining something like this, which wouldn't need the `sync.Once`:

```go
func (c *Client) InitRepo(inputRepo clients.Repo, commitSHA string, commitDepth int) error {
	// omitted
	repo, err := c.azdoClient.GetRepository(c.ctx, git.GetRepositoryArgs{
		Project:      &azdoRepo.project,
		RepositoryId: &azdoRepo.name,
	})
	// omitted error handling
	c.archived = *repo.IsDisabled
	...
}

func (c *Client) IsArchived() (bool, error) {
	return c.archived, nil
}
```

But if you stay with the `sync.Once` implementation, make sure to make a new one each time InitRepo is called.
```go
c.once = new(sync.Once)
```

File: clients/azuredevopsrepo/client.go
Comment (spencerschrock) [inline clients/azuredevopsrepo/client.go:140]: optional:
```go
return "", errDefaultBranchNotFound
```

**Review by spencerschrock — APPROVED** [review]
Just one small clarification to what I meant by "caching".

**Comment by justaugustus** [comment]
Wooo, thanks for this contribution, @JamieMagee!

**Pull request merged**
